### PR TITLE
graceful pod termination with preStop container life cycle hook

### PR DESCRIPTION
**Pull request by pontiyaraja**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
 /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**: Added new e2e test case, for graceful pod termination with preStop container life cycle hook

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
    added test case for the definition 
    `If the pod has defined a preStop hook, it is invoked inside of the pod. If the preStop hook is still running after the grace period expires, step 2 is then invoked with a small (2 second) extended grace period`
**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
/area conformance

### CODE REVIEW

**Comment by krmayankk** [comment]
/ok-to-test
Adding a link https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods where the above is being quoted for the test being added

File: test/e2e/node/pre_stop.go [hunk 1]
```diff
@@ -169,4 +178,105 @@ var _ = SIGDescribe("PreStop", func() {
...
+							PreStop: &v1.Handler{
+								Exec: &v1.ExecAction{
+									Command: []string{"sleep", "60"},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
```
Comment (krmayankk) [inline test/e2e/node/pre_stop.go]: there might be a helper for creating pod objects which you can use here

File: test/e2e/node/pre_stop.go [hunk 2]
```diff
@@ -169,4 +178,105 @@ var _ = SIGDescribe("PreStop", func() {
...
+		// save the running pod
+		pod, err = podClient.Get(pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "failed to GET scheduled pod")
+		framework.Logf("running pod: %#v", pod)
+
+		By("deleting the pod gracefully")
+		err = podClient.Delete(pod.Name, metav1.NewDeleteOptions(30))
+		Expect(err).NotTo(HaveOccurred(), "failed to delete pod")
+
+		By("verifying pod deletion was observed")
```
Comment (krmayankk) [inline test/e2e/node/pre_stop.go]: is it necessary to observe the watch event ?

File: test/e2e/node/pre_stop.go [hunk 3]
```diff
@@ -169,4 +178,105 @@ var _ = SIGDescribe("PreStop", func() {
...
+			}
+		}
+		if !deleted {
+			framework.Failf("Failed to observe pod deletion")
+		}
+
+		Expect(int64(deletedTime.Sub(lastPod.DeletionTimestamp.Time).Seconds())).To(SatisfyAll(
+			BeNumerically(">", *lastPod.DeletionGracePeriodSeconds),
+			BeNumerically("<", 75)))
+		Expect(lastPod.Spec.TerminationGracePeriodSeconds).ToNot(BeZero())
```
Comment (krmayankk) [inline test/e2e/node/pre_stop.go]: this is spec, why do we need to verify this ? We are giving the spec

File: test/e2e/node/pre_stop.go [hunk 4]
```diff
@@ -169,4 +178,105 @@ var _ = SIGDescribe("PreStop", func() {
...
+				framework.Failf("timed out waiting for pod deletion")
+			}
+		}
+		if !deleted {
+			framework.Failf("Failed to observe pod deletion")
+		}
+
+		Expect(int64(deletedTime.Sub(lastPod.DeletionTimestamp.Time).Seconds())).To(SatisfyAll(
+			BeNumerically(">", *lastPod.DeletionGracePeriodSeconds),
+			BeNumerically("<", 75)))
```
Comment (krmayankk) [inline test/e2e/node/pre_stop.go]: how do you verify that the preStop was indeed called ? By comparing the timestamp ? A comment explaining your assumptions will help maintain this test a lot better

**Comment by pontiyaraja** [comment]
cc @spiffxp

File: test/e2e/node/pre_stop.go [hunk 5]
```diff
@@ -169,4 +178,110 @@ var _ = SIGDescribe("PreStop", func() {
...
+					deleted = true
+				case watch.Error:
+					framework.Logf("received a watch error: %v", event.Object)
+					framework.Failf("watch closed with error")
+				}
+			case <-timer:
+				framework.Failf("timed out waiting for pod deletion")
+			}
+		}
+		if !deleted {
```
Comment (oomichi) [inline test/e2e/node/pre_stop.go]: nit: I guess this case never happens according to the above conditions which call Failf() or breaks the loop with deleted = true.

**Review by oomichi — COMMENTED** [review]
The part of this test seems to be copied from the existing "should be submitted and removed [Flaky]".
So it is nice to run this few times for verifying this is not flacky.

File: test/e2e/node/pre_stop.go [hunk 6]
```diff
@@ -169,4 +178,110 @@ var _ = SIGDescribe("PreStop", func() {
...
+
+		selector = labels.SelectorFromSet(labels.Set(map[string]string{"time": value}))
+		options = metav1.ListOptions{LabelSelector: selector.String()}
+		pods, err = podClient.List(options)
+		Expect(err).NotTo(HaveOccurred(), "failed to query for pods")
+		Expect(len(pods.Items)).To(Equal(0))
+	})
 })
+
+func createPodWithpreStopLifeCycle(name, value string) *v1.Pod {
```
[the remark this answers is not in the document]
Comment (pontiyaraja) [reply test/e2e/node/pre_stop.go]: I don't understand @oomicho

File: test/e2e/node/pre_stop.go (hunk 5 above)
Comment (pontiyaraja) [reply test/e2e/node/pre_stop.go]: Yes, this is a test case, so it should cover all.

File: test/e2e/node/pre_stop.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pontiyaraja) [reply test/e2e/node/pre_stop.go]: @oomichi

File: test/e2e/node/pre_stop.go [hunk 7]
```diff
@@ -169,4 +178,111 @@ var _ = SIGDescribe("PreStop", func() {
...
+		// We need to wait for the pod to be running, otherwise the deletion
+		// may be carried out immediately rather than gracefully.
+		framework.ExpectNoError(f.WaitForPodRunning(pod.Name))
+		// save the running pod
+		pod, err = podClient.Get(pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "failed to GET scheduled pod")
+		framework.Logf("running pod: %#v", pod)
+
+		By("deleting the pod gracefully")
+		err = podClient.Delete(pod.Name, metav1.NewDeleteOptions(30))
```
Comment (spiffxp) [inline test/e2e/node/pre_stop.go]: Why 30?

File: test/e2e/node/pre_stop.go [hunk 8]
```diff
@@ -169,4 +178,111 @@ var _ = SIGDescribe("PreStop", func() {
 	framework.ConformanceIt("should call prestop when killing a pod ", func() {
 		testPreStop(f.ClientSet, f.Namespace.Name)
 	})
+
+	It("graceful pod terminated should wait until preStop hook completes the process [Slow]", func() {
+		By("creating the pod")
+		name := "pod-prestop-hook-" + string(uuid.NewUUID())
+		value := strconv.Itoa(time.Now().Nanosecond())
+		pod := createPodWithpreStopLifeCycle(name, value)
```
Comment (spiffxp) [inline test/e2e/node/pre_stop.go]: nit: `podWithPreStopLifecycleHook`

To me, the "create" verb implies interaction with the cluster, whereas here we're just instantiating a test fixture

File: test/e2e/node/pre_stop.go [hunk 9]
```diff
@@ -169,4 +178,111 @@ var _ = SIGDescribe("PreStop", func() {
...
+		By("verifying pod deletion was observed")
+		deleted := false
+		var lastPod *v1.Pod
+		var deletedTime time.Time
+		timer := time.After(framework.DefaultPodDeletionTimeout)
+		for !deleted {
+			select {
+			case event, _ := <-w.ResultChan():
+				switch event.Type {
+				case watch.Deleted:
```
Comment (spiffxp) [inline test/e2e/node/pre_stop.go]: Depending on a specific event disqualifies this as a candidate for conformance tests

Depending on a watch means this test is following a known anti-pattern that can cause flakes (ref: https://github.com/kubernetes/kubernetes/issues/71646)

Can graceful termination be verified some other way?

**Review by spiffxp — COMMENTED** [review]
It's been a while since I've read e2e tests, so maybe some of this is a lack of familiarity/recency.

If this is intended to eventually be a conformance test, it's verifying behavior in a way that disqualifies it from conformance.  What's a user-visible way you could verify this without events or watches?

I feel like there are too many pre-checks in setup.  For example, instead of "query for pods, make sure they're 0, create a pod, query for pods, make sure there's 1, wait for the pod to go running" I would expect to see "create a pod and wait for it to go running"

File: test/e2e/node/pre_stop.go (hunk 9 above)
Comment (pontiyaraja) [reply test/e2e/node/pre_stop.go]: @spiffxp Can I use kubelet instead of watch ??

**Pull request merged**